encoding/ccf/decode_typedef.go

```diff
@@ -0,0 +1,477 @@
+		if err != nil {
+			return nil, err
+		}
+
+		// "Valid CCF Encoding Requirements" in CCF specs:
+		//
+		//   "field-name MUST be unique in composite-type."
+		//   "name MUST be unique in composite-type-value.fields."
+		if _, ok := fieldNames[field.Identifier]; ok {
+			return nil, fmt.Errorf("found duplicate field name %s in composite-type", field.Identifier)
+		}
+
+		// "Deterministic CCF Encoding Requirements" in CCF specs:
+		//
+		//   "composite-type.fields MUST be sorted by name"
+		//   "composite-type-value.fields MUST be sorted by name."
+		if !stringsAreSortedBytewise(previousFieldName, field.Identifier) {
+			return nil, fmt.Errorf("field names are not sorted in composite-type (%s, %s)", previousFieldName, field.Identifier)
+		}
+
+		fieldNames[field.Identifier] = struct{}{}
+		previousFieldName = field.Identifier
+		fields[i] = field
+	}
+
+	return fields, nil
+}
+
+// decodeCompositeField decodes field type as
+// language=CDDL
+//
+//	[
+//	    field-name: tstr,
+//	    field-type: inline-type
+//	]
+func (d *Decoder) decodeCompositeField(types *cadenceTypeByCCFTypeID, decodeTypeFn decodeTypeFn) (cadence.Field, error) {
+	// Decode array head of length 2
+	err := decodeCBORArrayWithKnownSize(d.dec, 2)
+	if err != nil {
+		return cadence.Field{}, err
+	}
+
+	// element 0: field-name
+	fieldName, err := d.dec.DecodeString()
+	if err != nil {
+		return cadence.Field{}, err
+	}
+
+	// element 1: field-type
+	fieldType, err := decodeTypeFn(types)
+	if err != nil {
+		return cadence.Field{}, err
+	}
+
+	if fieldType == nil {
+		return cadence.Field{}, errors.New("unexpected nil type as composite field type")
+	}
+
+	// Unmetered because decodeCompositeField is metered in decodeCompositeFields and called nowhere else
+	// fieldType is still metered.
+	return cadence.NewField(fieldName, fieldType), nil
+}
+
+// decodeInterfaceType decodes interface type as
+// language=CDDL
+// interface-type = [
+//
+//	id: id,
+//	cadence-type-id: tstr,
+//
+// ]
+func (d *Decoder) decodeInterfaceType(
+	types *cadenceTypeByCCFTypeID,
+	constructor func(common.Location, string) cadence.Type,
+) (ccfTypeID, cadenceTypeID, error) {
+
+	// Decode array head of length 2.
+	err := decodeCBORArrayWithKnownSize(d.dec, 2)
+	if err != nil {
+		return ccfTypeID(0), cadenceTypeID(""), err
+	}
+
+	// element 0: id
+	ccfID, err := d.decodeCCFTypeID()
+	if err != nil {
+		return ccfTypeID(0), cadenceTypeID(""), err
+	}
+
+	// "Valid CCF Encoding Requirements" in CCF specs:
+	//
+	//   "composite-type.id MUST be unique in ccf-typedef-message or ccf-typedef-and-value-message."
+	if types.has(ccfID) {
+		return ccfTypeID(0), cadenceTypeID(""), fmt.Errorf("found duplicate CCF type ID %d in interface-type", ccfID)
+	}
+
+	// element 1: cadence-type-id
+	cadenceID, location, identifier, err := d.decodeCadenceTypeID()
+	if err != nil {
+		return ccfTypeID(0), cadenceTypeID(""), err
+	}
+
+	_ = types.add(ccfID, constructor(location, identifier))
```

Comment:
```suggestion
	// The return value can be ignored, because its non-existence was already checked above
	_ = types.add(ccfID, constructor(location, identifier))
```

Reply:
Done in commit 19f829fd4ca3401237844db355ed247b6321d166.